code/datums/components/trapdoor.dm

```diff
@@ -0,0 +1,269 @@
+///distance a trapdoor will accept a link request.
+#define TRAPDOOR_LINKING_SEARCH_RANGE 7
+
+/**
+ * ## trapdoor component!
+ *
+ * component attached to floors to turn them into trapdoors, a constructable trap that when signalled drops people to the level below.
+ * assembly code at the bottom of this file
+ */
+/datum/component/trapdoor
+	///assembly tied to this trapdoor
+	var/obj/item/assembly/trapdoor/assembly
+	///path of the turf this should change into when the assembly is pulsed. needed for openspace trapdoors knowing what to turn back into
+	var/trapdoor_turf_path
+
+/datum/component/trapdoor/Initialize(starts_open, trapdoor_turf_path, assembly)
+	if(!isopenturf(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	src.assembly = assembly
+	if(IS_OPEN(parent))
+		openspace_trapdoor_setup(trapdoor_turf_path, assembly)
+	else
+		tile_trapdoor_setup(trapdoor_turf_path, assembly)
+
+	if(starts_open)
+		try_opening()
+
+///initializing as an opened trapdoor, we need to trust that we were given the data by a closed trapdoor
+/datum/component/trapdoor/proc/openspace_trapdoor_setup(trapdoor_turf_path)
+	src.trapdoor_turf_path = trapdoor_turf_path
+
+///initializing as a closed trapdoor, we need to take data from the tile we're on to give it to the open state to store
+/datum/component/trapdoor/proc/tile_trapdoor_setup(trapdoor_turf_path)
+	src.trapdoor_turf_path = parent.type
+	if(assembly && assembly.stored_decals.len)
+		reapply_all_decals()
+
+/datum/component/trapdoor/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_TURF_CHANGE, .proc/turf_changed_pre)
+	if(!src.assembly)
+		RegisterSignal(SSdcs, COMSIG_GLOB_TRAPDOOR_LINK, .proc/on_link_requested)
+	else
+		RegisterSignal(assembly, COMSIG_ASSEMBLY_PULSED, .proc/toggle_trapdoor)
+
+/datum/component/trapdoor/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(SSdcs, COMSIG_GLOB_TRAPDOOR_LINK)
+	UnregisterSignal(assembly, COMSIG_ASSEMBLY_PULSED)
+	UnregisterSignal(parent, COMSIG_TURF_CHANGE)
+
+/datum/component/trapdoor/proc/decal_detached(datum/source, description, cleanable, directional, pic)
+	SIGNAL_HANDLER
+	///so it adds the list to the list, not appending it to the end. thank you byond, very cool.
+	assembly.stored_decals += list(list(description, cleanable, directional, pic))
+
+/**
+ * ## reapply_all_decals
+ *
+ * changing turfs does not bring over decals, so we must perform a little bit of element reapplication.
+ */
+/datum/component/trapdoor/proc/reapply_all_decals()
+	for(var/list/element_data as anything in assembly.stored_decals)
+		apply_decal(element_data[1], element_data[2], element_data[3], element_data[4])
+	assembly.stored_decals = list()
+
+/// small proc that takes passed arguments and drops it into a new element
+/datum/component/trapdoor/proc/apply_decal(description, cleanable, directional, pic)
+	parent.AddElement(/datum/element/decal, _description = description, _cleanable = cleanable, _dir = directional, _pic = pic)
```

Comment:
this my friends is why we dont do underscore names and do the src trick instead

Reply:
this was so fucking painful holy mother of god I AM A BELIEVER NOW